skimage/filters/_multiotsu.pyx

```diff
@@ -0,0 +1,77 @@
+#cython: cdivision=True
+#cython: boundscheck=False
+#cython: nonecheck=False
+#cython: wraparound=False
+import numpy as np
+
+cimport numpy as cnp
+cimport cython
+from itertools import combinations
+
+def _find_threshold_multiotsu(double [:, ::1] var_btwcls,
+                              Py_ssize_t classes,
+                              Py_ssize_t bins):
+    """
+    Cython utility function for finding thresholds in multi-Otsu algorithm.
+    This function is only called by filters.threshold_multiotsu.
+
+    Parameters
+    ----------
+    var_btwcls : 2-d array
+        array of variance between classes
+    classes : int
+        number of classes to be segmented
+    bins : int
+        number of bins used in the histogram
+    """
+    cdef Py_ssize_t idd
+    cdef Py_ssize_t sh0, sh1
+    cdef double part_sigma = 0
+    # max_sigma is the maximum variance between classes.
+    cdef double max_sigma = 0
+
+    py_aux_thresh = np.empty(classes - 1, dtype=np.intp)
+    cdef Py_ssize_t[::1] aux_thresh = py_aux_thresh
+    cdef Py_ssize_t[::1] idx_tuple = np.zeros(classes+1, dtype=np.intp)
+    idx_tuple[classes] = bins - 1
+
+    with nogil:
+        _find_best_rec(var_btwcls=var_btwcls, min_val=0,
+                       max_val=bins-1, idx_tuple=idx_tuple,
+                       divisions=classes-1, depth=0, max_sigma=0,
+                       aux_thresh=aux_thresh)
+
+    return py_aux_thresh
+
+
+cdef double _find_best_rec(double[:, ::1] var_btwcls, cnp.intp_t min_val,
+                           cnp.intp_t max_val, Py_ssize_t[::1] idx_tuple,
+                           cnp.intp_t divisions, cnp.intp_t depth,
+                           double max_sigma, Py_ssize_t[::1] aux_thresh) nogil:
+    """
+    Recursive function for finding max_sigma.
```

Comment:
Recursion! Brilliant! =D

@alexandrejaguar even though it is a private function, would you mind adding a docstring explaining what it does and what all the arguments are? I find it very hard to read these long argument lists.